code/config/lib/Loot/Config/Generics.hs

```diff
@@ -0,0 +1,176 @@
+
+-- | Internal typeclass which splits 'Rec' into two specified pieces.
+class RSplit (l :: [k]) where
+    -- | The opposite to '<+>'.
+    rsplit :: Rec f (l ++ r) -> (Rec f l, Rec f r)
+
+instance RSplit '[] where
+    rsplit = (RNil, )
+
+instance RSplit ls => RSplit (l ': ls) where
+    rsplit (l :& xs) = let (ls, rs) = rsplit xs in (l :& ls, rs)
+
+-- | Internal typeclass which implements conversion between generic
+-- representation of a type and the respective config.
+class GDataIsoItem (x :: Type -> Type) where
+
+    -- | Configuration associated with given datatype.
+    type GFromData x :: [ItemKind]
+
+    -- | Converts datatype to config.
+    gToData :: KnownConfigKind k => x p -> ConfigRec k (GFromData x)
+
+    -- | Converts datatype from the associated config.
+    gFromData :: ConfigRec 'Final (GFromData x) -> x p
+
+instance GDataIsoItem x => GDataIsoItem (G.D1 i x) where
+    type GFromData (G.D1 i x) = GFromData x
+    gToData = gToData . G.unM1
+    gFromData = G.M1 . gFromData
+
+instance GDataIsoItem x => GDataIsoItem (G.C1 i x) where
+    type GFromData (G.C1 i x) = GFromData x
+    gToData = gToData . G.unM1
+    gFromData = G.M1 . gFromData
+
+instance GDataIsoItem (G.S1 ('G.MetaSel ('Just name) _1 _2 _3) (G.Rec0 a)) where
+    type GFromData (G.S1 ('G.MetaSel ('Just name) _1 _2 _3) (G.Rec0 a)) =
+        '[name ::: a]
+    gToData (G.M1 (G.K1 a)) = toItem a :& RNil
+    gFromData (ItemOptionF a :& RNil) = G.M1 (G.K1 a)
+
+instance TypeError ('Text "Datatype contains field without records") =>
+         GDataIsoItem (G.S1 ('G.MetaSel 'Nothing _1 _2 _3) x) where
+    type GFromData (G.S1 ('G.MetaSel 'Nothing _1 _2 _3) x) = '[]
+    gToData = error "impossible"
+    gFromData = error "impossible"
+
+instance ( GDataIsoItem x, GDataIsoItem y
+         , RSplit (GFromData x)
+         ) => GDataIsoItem (x :*: y) where
+    type GFromData (x :*: y) = GFromData x ++ GFromData y
+    gToData (x :*: y) = gToData x <+> gToData y
+    gFromData xs =
+      let (ls, rs) = rsplit @_ @(GFromData x) xs
+      in gFromData ls :*: gFromData rs
+
+instance GDataIsoItem G.U1 where
+    type GFromData G.U1 = '[]
+    gToData G.U1 = RNil
+    gFromData RNil = G.U1
+
+instance TypeError ('Text "Sum types are not yet supported") =>
+         GDataIsoItem (x :+: y) where
+    type GFromData (x :+: y) = '[]
+    gToData = error "impossible"
+    gFromData = error "impossible"
+
+instance TypeError ('Text "Empty variants configs are not allowed") =>
+         GDataIsoItem G.V1 where
+    type GFromData G.V1 = '[]
+    gToData = error "impossible"
+    gFromData = error "impossible"
+
+
+-- | Requires type to have a config associated with it.
+type DataIsoItem dt = (Generic dt, GDataIsoItem (G.Rep dt))
+
+-- | Derive configuration format corresponding to the given datatype.
+--
+-- Produced config is always flat: currently only product types are supported.
+-- Each field is represented in config as-is (no deep inspection is performed).
```

Comment:
Are these limitations fundamental or is it possible to support sum types and deep inspection? If it's possible, ideally an issue should be created with some thoughts on how to do it and a reference should be put here. However, since most likely nobody will do it anyway, that's really optional :smile:

Reply:
Mm, that's an interesting question. Probably, assuming that this generics functionality is ever useful, the use cases for sum types and nested product types will eventually appear.

I think I'll implement it here, but this will require a substantial rewrite.

Also, `FromData` (that is used like [here](https://github.com/serokell/lootbox/blob/01139a4215ab043d9bb47c6556277d61c79173fd/code/config/lib/Loot/Config/Generics.hs#L31)) will remain with the current limitations because we cannot just do a recursive descend over a nested datatype and assume every primitive type at the bottom to have `Generic` instance (instead, we can follow an existing configuration format to understand whether we are at a product type, a sum type or a primitive).

Reply:
> I think I'll implement it here, but this will require a substantial rewrite.

I am afraid that if we don't merge it now (≈this week), we won't merge this PR for another year at least :pled:

So I guess it's better to maybe write an additional sentence or create an issue and refer to it and merge this PR with existing limitations.

Reply:
Makes sense, I've added a line. This could be implemented in few days, but I'm not sure if this functionality would be used by anyone except me.